app/src/main/java/com/infomaniak/mail/ui/newMessage/NewMessageAiManager.kt

```diff
@@ -0,0 +1,246 @@
+            if (event == Lifecycle.Event.ON_DESTROY) {
+                _aiViewModel = null
+                valueAnimator?.cancel()
+                valueAnimator = null
+            }
+        })
+    }
+
+    fun observeEverything() {
+        observeAiOutput()
+        observeAiPromptStatus()
+        observeAiFeatureFlagUpdates()
+    }
+
+    private fun observeAiOutput() = with(binding) {
+        aiViewModel.aiOutputToInsert.observe(viewLifecycleOwner) { (subject, content) ->
+            subject?.let { subjectTextField.setText(it) }
+            bodyText.setText(content)
+        }
+    }
+
+    private fun observeAiPromptStatus() {
+        aiViewModel.aiPromptOpeningStatus.observe(viewLifecycleOwner) { (shouldDisplay, shouldResetContent, becauseOfGeneration) ->
+            if (shouldDisplay) onAiPromptOpened(shouldResetContent) else onAiPromptClosed(becauseOfGeneration)
+        }
+    }
+
+    private fun onAiPromptOpened(resetPrompt: Boolean) = with(binding) {
+        if (resetPrompt) {
+            aiViewModel.apply {
+                aiPrompt = ""
+                aiPromptOpeningStatus.value?.shouldResetPrompt = false
+            }
+        }
+
+        // Keyboard is opened inside onCreate() of AiPromptFragment
+
+        val foundFragment = childFragmentManager.findFragmentByTag(AI_PROMPT_FRAGMENT_TAG) as? AiPromptFragment
+        if (aiPromptFragment == null) {
+            aiPromptFragment = foundFragment ?: AiPromptFragment()
+        }
+
+        if (foundFragment == null) {
+            childFragmentManager
+                .beginTransaction()
+                .add(aiPromptFragmentContainer.id, aiPromptFragment!!, AI_PROMPT_FRAGMENT_TAG)
+                .commitNow()
+        }
+
+        scrim.apply {
+            isVisible = true
+            isClickable = true
+        }
+
+        animateAiPrompt(true)
+        setAiPromptVisibility(true)
+        newMessageConstraintLayout.descendantFocusability = FOCUS_BLOCK_DESCENDANTS
+    }
+
+    private fun onAiPromptClosed(withoutTransition: Boolean) = with(binding) {
+
+        fun removeFragmentAndHideScrim() {
+            aiPromptFragment?.let {
+                childFragmentManager
+                    .beginTransaction()
+                    .remove(it)
+                    .commitNow()
+            }
+            aiPromptFragment = null
+
+            scrim.isGone = true
+        }
+
+        aiPromptFragmentContainer.hideKeyboard()
+
+        if (withoutTransition) {
+            removeFragmentAndHideScrim()
+        } else {
+            viewLifecycleOwner.lifecycleScope.launch {
+                delay(animationDuration)
+                removeFragmentAndHideScrim()
+            }
+        }
+
+        if (!withoutTransition) animateAiPrompt(false)
+        setAiPromptVisibility(false)
+        newMessageConstraintLayout.descendantFocusability = FOCUS_BEFORE_DESCENDANTS
+    }
```

Comment:
We can avoid that inner function by doing this maybe:
``` kotlin
private fun onAiPromptClosed(withoutTransition: Boolean) = with(binding) {
        aiPromptFragmentContainer.hideKeyboard()
        viewLifecycleOwner.lifecycleScope.launch {
            if (withoutTransition) delay(animationDuration)
            childFragmentManager.beginTransaction().remove(aiPromptFragment).commitNow()
            scrim.isGone = true
        }

        if (!withoutTransition) animateAiPrompt(false)
        setAiPromptVisibility(false)
        newMessageConstraintLayout.descendantFocusability = FOCUS_BEFORE_DESCENDANTS
    }
```

Reply:
Could there be an advantage in avoiding to launch a coroutine if we don't need it? Because else, that's a nice idea. 

If we do get rid of the double function call, I still like the idea of have the lines wrapped inside a more meaningful function like `removeFragmentAndHideScrim()` though, what do you think?